### PR TITLE
Modernize the Stack Kotlin Edition

### PR DESCRIPTION
**Pull request by fwiesel**
Courtesy of the context menu function "Convert Java File to Kotlin File..." of Android Studio, plus some minor adjustments.

It works at least in my short tests even on older android phones (virtual), which surprises me, as the version of Java is much newer than I expected to work.
I wasted some time to switch it to an older version.
I tested it on my current one against the published version on another phone shortly.

I have no experience in kotlin, so I can't say if the code is idiomatic.

### CODE REVIEW

File: app/src/main/java/de/rochefort/childmonitor/AudioCodecDefines.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/AudioCodecDefines.kt:22]: This should not be a class anymore, it can just be an object.
(It could also just be a kt file, but I think it's worth keeping the String AudioCodecDefines around)

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt:2]: License comment needs to be adjusted

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt:47]: this was already unused before, lets get rid of it

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt:65]: this looks weird. And android studio complains about it. Why run?

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:36]: If we are going to use a global TAG it should somewhere else. If it is used in DiscoverActivity only we should move it to a companion object

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:98]: using !! here breaks the null safety that kotlin provides. We should avoid this operator like the plague.
Rather use the ? operator, possibly in combination with let{}
On a side note: Many things that we just changed to local vars for null checking can now be done more elegantly using ?

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:111]: don't use Runnables in Java. Use () -> Unit instead

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:180]: One this is no longer a Runnable you'll be able to just call it, instead of calling run()

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:238]: this could should probably never have gone into the getter but into the constructor instead

File: app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt:111]: findViewById does not promise to return non null value, so maybe we should use the ? operator here?

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/ListenService.kt:40]: explicitly adding companion here is not needed

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/ListenService.kt:62]: use intent.extras?.let instead of the null check here

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/ListenService.kt:137]: Unrelated to kotlin, but can we get rid of those m prefixes? 

also here don't use Runnables

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/java/de/rochefort/childmonitor/ListenService.kt:161]: It's unidiomatic to throw Exceptions in Kotlin.

See e.g. https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07

**Review by enguerrand — CHANGES_REQUESTED** [review]
Thanks for doing this. Unfortunately the automatic kotlin conversion stuff does not always produce idiomatic kotlin, at least in my experience.

I have added some initial feedback, there surely is a lot more to change so I guess to get this right this will be a larger undertaking. Feel free to go back to the original PR if you prefer.

Disclaimer: I have never used kotlin in Android, only general kotlin, so I might get things wrong here and there.

That being said, some more general feedback:
- kt files belong into a different source root (src/main/kotlin)
- Gradle sync fails on my machine saying: "The project is using an incompatible version (AGP 8.2.2) of the Android Gradle plugin. Latest supported version is AGP 8.0.2"
- Moving forward, if you want to go ahead with this, please make adjustments in additional commits rather than rebasing / force pushing.

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:98]: I suspect the reason it is done that way, that in fact it can never be null, because of the activity life-cycle.

`nsdManager` is set to a non-null value after `onCreate`. Probably the clean way to solve that is to extract that code to some class that is created in `onCreate`.

File: app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt:111]: While the function does not guarantee not to return null, it would only return null, if the resources do not match the code. And then the question is, how to handle that?
I think, simply crashing with a nice stack-trace would be preferable over continuing to work but having some fields unset.

Probably, the pattern in the android documentation should be followed:
https://developer.android.com/kotlin/common-patterns#fragment-nullability

File: app/src/main/java/de/rochefort/childmonitor/AudioCodecDefines.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/AudioCodecDefines.kt:22]: Sure thing.

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt:47]: Will do.

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt:2]: In which way?

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:36]: Will do

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:180]: It turns into `multicastReleaser?.invoke()`  😄

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt:238]: Moved it to an `init()`

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt:40]: Gone with the move to a simple object.

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt:62]: Done (and other places where I did use that pattern).

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt:137]: Done both.

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt]: I've wrapped the library functions that throw exceptions in `runCatching` blocks.

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt]: From what I read, the standard practice is apparently to mark it as a `lateinit` variable.

File: app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt]: I agree with the logic that if this returns null we are looking at a bug so it's okay to crash.
Lets leave it as is, the lateinit stuff does not apply here since this is a local variable.

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt]: This was the comment from the file I copied originally, which was a java file. Since it is now changed quite drastically, I am assuming that we needed to somehow state that the current code is based on a state that is subject to the above license statement, but was now converted to kotlin.
Probably best below my "Taken from" notice.

File: app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt]: Yep, agreed, this makes sense here.
I am generally reluctant to use lateinit because in imho it is just another way to basically ignore the nullability of a field that make everything crash if the programmer's assumptions are wrong, but I agree with your lifecycle rationale.

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt]: I've added a short notice.
I'd say, that without different notice, the code and the changes are subject to the Apache 2.0 license.

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt]: I personally find the runCatching stuff in combination with result.isFailure more clumsy than try - catch.
Therefore  I don't mind having try catch blocks when calling external api that throws. I am just saying that we should just not propagate exceptions that are part of normal runtime situations. Instead, we should return a value from which we the caller can derive error states.
The behavior is changed now in that streamAudio only returns instead of propagating error state to the caller. This is mostly okay in this particular case because the caller does not need a return value, but we are still losing the "connection failed" logging in the current code state

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt:75]: It looks a bit weird to access this.listenThread in the lambda.
I think in this case I'd do this:
```
this.listenThread?.interrupt()
this.listenThread = null     
```

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt:166]: Not sure about the purpose of this method.

Also the return value is pointless (and unused).

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:171]: We can get rid of this temp variable with another let block.

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:240]: also here: move this line out of the lambda. it is always okay to set this to null.

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:56]:
```suggestion
            volumeHistory[size - 1].coerceAtLeast(0.3)
```

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt:81]: this should not be all caps

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:178]:
```suggestion
                    portText.text = port.toString()
```

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:76]:
```suggestion
        val leftMost = (volumeHistory.size() - width).coerceAtLeast(0)
```

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:80]:
```suggestion
        val length = min(size, width)
```

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:89]: explicit getter/setter functions are unidiomatic in kotlin

File: settings.gradle
Comment (enguerrand) [inline settings.gradle:2]: Can you help me understand which problem exactly we are solving by introducing this?

**Review by enguerrand — CHANGES_REQUESTED** [review]
It is starting to look quite nice, we are getting close... Thanks once more for your efforts!

I would help me if you could let me resolve the discussions. This way I can track more easily which parts I already reviewed.

File: app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt]: I always struggle with licensing questions. In this case I think that neither the kotlin conversion nor the method removal justifies any copyright claims on the resulting code. I would hence not add my own license statement and leave it licensed under the apache license.

I am still unclear about the exact wording I'd like to have in the additional note though. 

How about this:
```
 *
 * Originally taken from https://android.googlesource.com/platform/external/nist-sip/+/6f95fdeab4481188b6260041b41d1db12b101266/src/com/android/sip/media/G711UCodec.java
 * Modifications to the original code:
 *   - Source conversion to kotlin
 *   - Removal of unused method getSampleCount
```

Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt]: Well, it's your project :). I wouldn't make any claims, but the license would allow for that.

My understanding is, it is sufficient to say that it has been modified:
> You must cause any modified files to carry prominent notices stating that You changed the files;

So no need to keep a change-log. It is only to avoid confusion.

I.e.:
```
 * Originally taken from https://android.googlesource.com/platform/external/nist-sip/+/6f95fdeab4481188b6260041b41d1db12b101266/src/com/android/sip/media/G711UCodec.java
 * Adopted to the needs of the Child Monitor project.
```

The example from an Apache NOTICE file as [reference](https://infra.apache.org/licensing-howto.html#example-notice).

But it is your call, of course.

Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/audio/G711UCodec.kt]: I like your proposal, let's go with this!

File: app/src/main/kotlin/de/rochefort/childmonitor/DiscoverActivity.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/DiscoverActivity.kt:161]: Will be fixed with the next push.

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt:75]: Will be fixed with the next push.

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:171]: Done

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:240]: Done.

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:56]: Done

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt:81]: Passed simply `R.string.listening` directly.

File: app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt:178]: Done

File: settings.gradle
Comment (fwiesel) [reply settings.gradle:2]: Sure, that ensures that if the toolchain needed for the project is missing, then it gets installed.
See here for the details: https://docs.gradle.org/current/userguide/toolchains.html#sub:download_repositories
I didn't have the right one for some reason, and that stanza was solving that.

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: It sets up the audio-track for the passed block, and closes and stops the track in any case.

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt]: Made it public.

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: I meant the purpose of extracting a method for this. 

I guess it's a matter of taste personally I don't find the code to be more readable like this.

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [inline app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt:29]: This public is redundant

File: settings.gradle
Comment (enguerrand) [reply settings.gradle:2]: Ok, sounds reasonable

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (enguerrand) [reply app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt]: The public is redundant, members are public if they are not declared private or internal.
 (as pointed out in a different comment already

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt]: > I personally find the runCatching stuff in combination with result.isFailure more clumsy than try - catch.

I agree, I misunderstood you then.

> This is mostly okay in this particular case because the caller does not need a return value, but we are still losing the "connection failed" logging in the current code state

Sure, but what are you going to do with that information?
On top of that, is it an error, when I close the monitor on the child device by stopping the child activity?
It closes the connection though.

I've re-introduced the log  message though.

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/VolumeView.kt]: I've removed it.

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: I am  coming from C++ / C side, and always concerned about keeping the open / close, acquire / release, or in this case startRecording / stop methods close together, and ensure that they are always called.

I was thinking, I could write also something like
```
        ... setup audioTrack ...
        Closeable { audioTrack.stop() }.use {
          ...
        }
```
but the we are like four levels of indent down, and the use is a block itself.

File: app/src/main/java/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [reply app/src/main/java/de/rochefort/childmonitor/ListenService.kt]: The point I was trying to make is that your refactoring has changed the error handling logic.
It's okay to do that, but it should be in a separate step and not as part of a refactoring.
This is useful when tracking down bugs using git bisect.
Let me give you an example of changed error handling:
Previously, if the remote address was unreachable, the instantiation of the  would have thrown a SocketException, which in turn would have been caught by a try catch within the Thread block:
```
        val lt = Thread {
            try {
                val socket = Socket(address, port)
                streamAudio(socket)
            } catch (e: IOException) {
                Log.e(TAG, "Failed to stream audio", e)   // <-- here!
            }
```
Now that the handling of that error is no longer defined.
The change I was looking for when I made the original comment was a refactoring that returns Results where Exceptions were thrown previously, and handles failed results where exceptions would previously have been caught.
If it is still unclear, feel free to just revert the changes to the exception handling and I will fix it up to my taste after merging.

File: app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt
Comment (enguerrand) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: > keeping the open / close, acquire / release, or in this case startRecording / stop methods close together

Okay, I can agree with that logic, but then again, such a refactoring should not change semantics _in the refactoring step_
If the semantics need to be improved this should happen in a different commit.

Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: I think, I've undone the refactoring.

Comment (enguerrand) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: Unfortunately still not... The current state still suffers from the problem I outlined in [this post](https://github.com/enguerrand/child-monitor/pull/12#discussion_r1509561136)

The Socket constructor can throw an Exception which was caught in the original code and is no longer caught now.

Comment (fwiesel) [reply app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt]: True, I missed that. I hope, I have handled it now in an acceptable manner.

**Comment by enguerrand** [comment]
Looks good. Thanks (once more) for this contribution!

**Pull request merged**
